source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -434,6 +445,22 @@ ReadOrParseState Filter::readProxyHeader(Network::IoHandle& io_handle) {
     if (nread < 1) {
       ENVOY_LOG(debug, "failed to read proxy protocol (no bytes read)");
       return ReadOrParseState::Error;
+    } else if (config_.get()->allowRequestsWithoutProxyProtocol() && buf_off_ == 0) {
+
+      if (nread < PROXY_PROTO_V1_SIGNATURE_LEN) {
+        ENVOY_LOG(debug, "request does not have enough bytes to determine if v1 or v2 proxy "
+                         "protocol, forwarding as is");
+        return ReadOrParseState::SkipFilter;
+      }
+
+      if ((nread < PROXY_PROTO_V2_SIGNATURE_LEN ||
+           memcmp(buf_, PROXY_PROTO_V2_SIGNATURE, PROXY_PROTO_V2_SIGNATURE_LEN)) &&
+          memcmp(buf_, PROXY_PROTO_V1_SIGNATURE, PROXY_PROTO_V1_SIGNATURE_LEN)) {
```

Comment:
it should be
`((nread < PROXY_PROTO_V2_SIGNATURE_LEN ||
           memcmp(buf_, PROXY_PROTO_V2_SIGNATURE, nread)) &&
          memcmp(buf_, PROXY_PROTO_V1_SIGNATURE, nread))`

Reply:
this is actually fine as is. we already know at this point (due to short-circuiting in this if statement and the check above) that `nread` is larger than both `PROXY_PROTO_V2_SIGNATURE_LEN` and `PROXY_PROTO_V1_SIGNATURE_LEN`. We don't want to check more bytes than needed in `buf_` to check for either signature.

Reply:
we actually only need to update the second check (v1 check), since for both we want the min of `nread` and the signature length. but we know `nread` is >= `PROXY_PROTO_V2_SIGNATURE_LEN` after short-circuiting in the initial `if`. addressed the latter check in https://github.com/envoyproxy/envoy/pull/18951/commits/b5cbfc18d12f9f7fd61cf68467cf78612014fcd9

Reply:
this is actually fine as is. the first if statement means we know `nread` >= `PROXY_PROTO_V1_SIGNATURE_LEN`. the in this if statement the `nread < PROXY_PROTO_V2_SIGNATURE_LEN` case short-circuits if needed, so in the second `memcmp` we know `nread` >= `PROXY_PROTO_V2_SIGNATURE_LEN`.

if it helps with readability I'm happy to add `std::min` on `nread` and the signature lengths, but I do not think there is a logic error here

Reply:
sorry, you are right. 

I was based on the logic we want to exam more bytes.